### PR TITLE
[SPARK-24123][SQL] Fix precision issues in monthsBetween with more than 8 digits

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

SPARK-23902 introduced the ability to retrieve more than 8 digits in `monthsBetween`. Unfortunately, current implementation can cause precision loss in such a case. This was causing also a flaky UT.

This PR mirrors Hive's implementation in order to avoid precision loss also when more than 8 digits are returned.

## How was this patch tested?

running 10000000 times the flaky UT

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @ueshin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 1]
```diff
@@ -45,6 +45,7 @@ object DateTimeUtils {
   // it's 2440587.5, rounding up to compatible with Hive
   final val JULIAN_DAY_OF_EPOCH = 2440588
   final val SECONDS_PER_DAY = 60 * 60 * 24L
+  final val SECONDS_PER_MONTH = 60 * 60 * 24 * 31D
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Thank you for working on this, @mgaido91 !
BTW, is the naming, `SECONDS_PER_MONTH`, correct for February?

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: hi @dongjoon-hyun! thanks for the comment. Of course it is not, strictly speaking, but this is how both Spark before the PR (notice the 31.0) and Hive (which Spark draws his inspiration from for this function) work. They consider every month as having 31 days, even though this is not true of course.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I see now your point, yes, I will inline the value as Hive does with a comment then, thanks.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
SparkR failure is irrelevant to this PR and occurs in several other PRs. We had better wait for fixing SparkR.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 2]
```diff
@@ -888,14 +888,19 @@ object DateTimeUtils {
...
-      return (months1 - months2).toDouble
+      return monthDiff
     }
-    // milliseconds is enough for 8 digits precision on the right side
-    val timeInDay1 = millis1 - daysToMillis(date1, timeZone)
-    val timeInDay2 = millis2 - daysToMillis(date2, timeZone)
-    val timesBetween = (timeInDay1 - timeInDay2).toDouble / MILLIS_PER_DAY
-    val diff = (months1 - months2).toDouble + (dayInMonth1 - dayInMonth2 + timesBetween) / 31.0
+    // using milliseconds can cause precision loss with more than 8 digits
+    // we follow Hive's implementation which uses seconds
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I checked how Hive works. It works as this comment says.

**Comment by gatorsmile** [comment]
The R test failure is not related. Thanks! Merged to master

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 2 above)
Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: LGTM

**Comment by dongjoon-hyun** [comment]
Thank you, @gatorsmile and @mgaido91 !

**Pull request closed without merging**